core/trino-main/src/main/java/io/trino/execution/scheduler/HashDistributionSplitAssigner.java

```diff
@@ -183,20 +230,44 @@ private static Map<Integer, TaskPartition> createOutputPartitionToTaskPartition(
         OutputDataSizeEstimate mergedEstimate = OutputDataSizeEstimate.merge(partitionedSourcesEstimates);
         ImmutableMap.Builder<Integer, TaskPartition> result = ImmutableMap.builder();
         PriorityQueue<PartitionAssignment> assignments = new PriorityQueue<>();
-        assignments.add(new PartitionAssignment(new TaskPartition(), 0));
-        for (int outputPartitionId = 0; outputPartitionId < partitionCount; outputPartitionId++) {
-            long outputPartitionSize = mergedEstimate.getPartitionSizeInBytes(outputPartitionId);
-            if (assignments.peek().assignedDataSizeInBytes() + outputPartitionSize > targetPartitionSizeInBytes
-                    && assignments.size() < partitionCount) {
-                assignments.add(new PartitionAssignment(new TaskPartition(), 0));
+        for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+            long partitionSizeInBytes = mergedEstimate.getPartitionSizeInBytes(partitionId);
+            if (assignments.isEmpty() || assignments.peek().assignedDataSizeInBytes() + partitionSizeInBytes > targetPartitionSizeInBytes || !canMerge) {
+                Optional<PlanNodeId> splitBy = Optional.empty();
+                int subPartitionCount = 1;
+                if (partitionSizeInBytes > targetPartitionSizeInBytes) {
+                    Map<PlanNodeId, Long> sourceSizes = getSourceSizes(partitionedSources, outputDataSizeEstimates, partitionId);
+                    PlanNodeId largestSource = sourceSizes.entrySet().stream()
+                            .max(Map.Entry.comparingByValue())
+                            .map(Map.Entry::getKey)
+                            .orElseThrow();
+                    long largestSourceSizeInBytes = sourceSizes.get(largestSource);
+                    long remainingSourcesSizeInBytes = partitionSizeInBytes - largestSourceSizeInBytes;
+                    if (remainingSourcesSizeInBytes <= targetPartitionSizeInBytes / 4 && canSplit.test(largestSource)) {
+                        long targetLargestSourceSizeInBytes = targetPartitionSizeInBytes - remainingSourcesSizeInBytes;
+                        splitBy = Optional.of(largestSource);
+                        subPartitionCount = toIntExact(largestSourceSizeInBytes / targetLargestSourceSizeInBytes) + 1;
```

Comment:
show we put some limit on how many sub partitions can we generate - I do not see a strict reason - but I may be missing sth

Reply:
Currently we don't set any arbitrary limit for arbitrary distributed tasks. Here in theory it should be similar (it is in theory an "arbitrary distribution" but just within only a single partition)